<h3>Upload Instructions</h3>
<p>We recommend uploading two separate files for phenotype and genotype data through dataverse.</p>

<h4>Genotype data files</h4>
<p>Accepted formats are FASTQ or VCF. If you upload, a VCF please describe or indicate:</p>
    <ol>
        <li>How your VCF was generated</li>
        <li>What software was used and which thresholds were applied</li>
        <li>If the variants are filtered and how this was performed</li>
    </ol>
<h4>Phenotype data files</h4>
<ul>
    <li>We recommend formatting your file as a tab delimited or a comma delimited file. Excel files are also acceptable.</li>
    <li>Please include an ID field that links directly to the names of the FASTQ or VCF files that you upload.</li>
    <li>Drug resistance results should include information on what concentration of the drug was tested. For example if isoniazid drug resistance was tested at a concentration of 0.2 then the field containing the resistance status should be called inh_0.2 and so forth for other drugs. Please indicate the concentration units and method of testing in the description of your file during the dataverse upload.</li>
    <li>Please include data on when, month &amp; year, and where the sample was collected including city &amp; country. If GIS or longitude/latitude coordinates are available please include these as well. Please do not list postal or zip codes due to privacy concerns.</li>
    <li>Adding data on traditional fingerprint (RFLP, MIRU-VNTR, Spoligotype) is also helpful, and any completely de-identified data on the host is also acceptable.</li>
</ul>
