{% extends "base.html" %}
{% load staticfiles %}

{% block title %}
Predict - genTB
{% endblock %}


{% block content %}

        <div class="row">
            <div class="col-lg-12">
                <h1 class="page-header">Predict
                <small>Upload your genetic data to make drug resistance prediction.
                </small><h1>
            </div>
        </div>

        <div class = "row">
            <div class="col-md-6 portfolio-item">
                {% if PREVIOUS_ATTEMPT_DELETED %}
                    <div class="alert alert-success" role="alert">
                        Note: Your previous dataset information, including the dropbox link, was removed.
                    </div>
                {% endif %}
                <h3>Upload your genetic data as:</h3>
                    <ul style="font-size: 140%;">
                        <li>a FASTQ file or</li>
                        <li>a variant call file (vcf file)</li>
                    </ul>
                <p>To examine the frequency of individual mutations by drug resistance you can consult the <a href="{% url 'view_map_page' %}">Map</a> function for country specific data, and the <a href="{% url 'view_explore_page' %}">Explore</a> function to examine these frequency in all of the data.</p>

                {% if not IS_LOGGED_IN %}


                <div class="panel panel-info">
                  <div class="panel-body">
                    <p>To contribute a file for prediction, please:
                        {% include "login_signup_links.html" %}
                    </p>
                  </div>
                </div>

                {% else %}
                    <h4><i>Upload VCF or FASTQ Files:</i></h4>



                    {% include "predict/predict_form.html" %}
               {% endif %}

            </div>

            <div class="col-md-6 portfolio-item">
<h3>vcf file</h3>
<p>VCF stands for Variant Call Format and was used by the 1000 Genomes project to encode SNPs and other structural genetic variants. The format is currently maintained by <a href="http://ga4gh.org/#/fileformats-team">The Global Alliance Data Working Group File Formats Task Team</a>. This is further described <a href="http://samtools.github.io/hts-specs/VCFv4.2.pdf">here</a>.</p>

<p>Minimal genotypic information for accurate resistance predictions: genetic regions are listed in order of decreasing importance: (for more detailed list of genetic variants see reference Farhat MR, Sultana R et al. under review)</p>

    <table class="table">
       <tr>
           <th align = "center">Drug</th>
           <th align = "cetner">Genes</th>
       </tr>
       <tr>
           <td>Isoniazid</td>
           <td><i>katG</i>, <i>inhA</i> promoter, <i>inhA</i>, <i>embB</i>,
               <i>iniB</i>, <i>kasA</i>, <i>embAB</i> promoter, and <i>ahpC</i>
           </td>
       </tr>
       <tr>
           <td>Rifampicin</td>
           <td><i>rpoB</i></td>
       </tr>
       <tr>
           <td>Pyrazinamide</td>
           <td><i>pncA</i> and <i>pncA</i> promoter</td>
       </tr>
       <tr>
           <td>Ethambutol</td>
           <td><i>embB</i>, <i>iniB</i>, <i>embAB</i> promoter, <i>embAB</i>, and <i>embC</i></td>
       </tr>
       <tr>
           <td>Streptomycin</td>
           <td><i>rpsL</i>, <i>rrs</i>, <i>gid</i>, and <i>rrs-rrl</i> intergenic region</td>
       </tr>
       <tr>
           <td>Ethionamide</td>
           <td><i>inhA promoter</i>, <i>ethA</i>, and <i>inhA</i></td>
       </tr>
       <tr>
           <td>Kanamycin*</td>
           <td><i>rrs</i>, <i>tlyA</i>*</td>
       </tr>
       <tr>
           <td>Capreomycin</td>
           <td><i>rrs</i></td>
       </tr>
       <tr>
           <td>Amikacin</td>
           <td><i>rrs</i></td>
       </tr>
       <tr>
           <td>Fluoroquinolones</td>
           <td><i>gyrA</i>, <i>gyrB</i></td>
       </tr>
       <tr>
           <td>Para-aminosalicylic acid</td>
           <td><i>thyA</i></td>
       </tr>
       </table>
       * Note: Gene <i>eis</i> is a predictor of kanamycin resistance, but data for this gene was not available when the predictive model was built.
        </div>
    </div>


        <div class="col-md-6" id="d3-target-div" style="padding-left:60px;">
        </div>
        <script type="text/javascript" src="{{ STATIC_URL }}{% static 'd3/d3.v3.min.js' %}"></script>
        <script type="text/javascript" src="{{ STATIC_URL }}js/app.js"></script>
{% endblock %}
